### PR TITLE
Add the `-via` arguments to config file and GUI for the viewer

### PR DESCRIPTION
**Pull request by dmcmahill**
`vncviewer` already includes a `-via <ssh_host>` command line option to
specify that an SSH tunnel to `<ssh_host>` should be established and
VNC traffic routed through the tunnel.  However, that option was not
available either in the `*.tigervnc` setup files or via the GUI.

This adds `UseSSH` (boolean) and `via` (string) parameters to the config
file and exposes them to the VNC viewer GUI under the options dialog.

This shows where I added the option to the GUI:

tigervnc_via

This shows an example of how the option shows up in a `.tigervnc` file:
```
UseSSH=1
via=my.sshhost.com
```

### CODE REVIEW

File: vncviewer/OptionsDialog.cxx [hunk 1]
```diff
@@ -58,7 +58,7 @@ std::map<OptionsCallback*, void*> OptionsDialog::callbacks;
 static std::set<OptionsDialog *> instances;
 
 OptionsDialog::OptionsDialog()
-  : Fl_Window(450, 460, _("VNC Viewer: Connection Options"))
+  : Fl_Window(450, 550, _("VNC Viewer: Connection Options"))
```
Comment (CendioOssman) [inline vncviewer/OptionsDialog.cxx:61]: I would prefer to not keep growing this dialog. I wonder if the new dialog from #1410 should be merged first to solve this.

File: vncviewer/parameters.cxx [hunk 2]
```diff
@@ -158,12 +158,14 @@ StringParameter display("display",
...
                         "F8");
 
+
 BoolParameter fullscreenSystemKeys("FullscreenSystemKeys",
                                    "Pass special keys (like Alt+Tab) directly "
                                    "to the server when in full-screen mode.",
                                    true);
 
 #ifndef WIN32
+BoolParameter useSSH("UseSSH", "Create SSH tunnel", false);
```
Comment (CendioOssman) [inline vncviewer/parameters.cxx:168]: Why this new parameter? Isn't it sufficient to leave the via setting empty to disable the feature?

File: vncviewer/OptionsDialog.cxx [hunk 3]
```diff
@@ -712,6 +721,38 @@ void OptionsDialog::createSecurityPage(int tx, int ty, int tw, int th)
   tx = orig_tx;
   ty += INNER_MARGIN;
 
+    /* Connection */
```
Comment (CendioOssman) [inline vncviewer/OptionsDialog.cxx:724]: This feature doesn't exist on Windows, so there needs to be an `#ifdef` around these things.

**Review by CendioOssman — CHANGES_REQUESTED** [review]
Nice! This would make this feature more accessible to others. I have some small concerns, though. See the comments.

File: vncviewer/OptionsDialog.cxx (hunk 1 above)
Comment (dmcmahill) [reply vncviewer/OptionsDialog.cxx:61]: I'm just back at the computer after being AFK for a week.  I took a look at #1410 and that looks pretty nice.  Do you have a sense for when that may merge?  I'm happy to rework my changes to use the new dialog.

File: vncviewer/parameters.cxx (hunk 2 above)
Comment (dmcmahill) [reply vncviewer/parameters.cxx:168]: I think you're right.  I'll remove this parameter and just check if the host is a blank string or not.

File: vncviewer/OptionsDialog.cxx (hunk 1 above)
Comment (CendioOssman) [reply vncviewer/OptionsDialog.cxx:61]: #1410 has spiralled a bit out of control, but I can try to get the option window part merged so it doesn't block the work here.

**Comment by dmcmahill** [comment]
I've reworked the code from this PR into a branch based on the work in https://github.com/TigerVNC/tigervnc/pull/1529.  Once that PR merges I'll rebase the branch and resubmit this PR using the new code.  In the mean time, my new branch is at https://github.com/dmcmahill/tigervnc/commits/newoptions-via-to-gui

**Comment by CendioOssman** [comment]
Sounds good. But please don't open a new PR. Simply rebase and force push an update to the PR here. That way we keep all history.

**Comment by dmcmahill** [comment]
> Sounds good. But please don't open a new PR. Simply rebase and force push an update to the PR here. That way we keep all history.

Understood.  I went ahead and force pushed the new code that is still rooted in your branch so I'll need to do one more rebase and force push once the other PR merges.  But at least this PR now shows what the change will be.

Here is the options dialog with the new entry.

tigervnc_via_newoptions

**Comment by CendioOssman** [comment]
One goal of the new layout is to avoid having to increase the size of the options dialog, and instead have scroll bars. My idea was something like this:

```diff
diff --git a/vncviewer/OptionsDialog.cxx b/vncviewer/OptionsDialog.cxx
index 5550ade7..dd295abe 100644
--- a/vncviewer/OptionsDialog.cxx
+++ b/vncviewer/OptionsDialog.cxx
@@ -53,6 +53,7 @@
 #include <FL/Fl_Round_Button.H>
 #include <FL/Fl_Int_Input.H>
 #include <FL/Fl_Choice.H>
+#include <FL/Fl_Scroll.H>
 
 using namespace std;
 using namespace rdr;
@@ -63,7 +64,7 @@ std::map<OptionsCallback*, void*> OptionsDialog::callbacks;
 static std::set<OptionsDialog *> instances;
 
 OptionsDialog::OptionsDialog()
-  : Fl_Window(580, 480, _("TigerVNC Options"))
+  : Fl_Window(580, 420, _("TigerVNC Options"))
 {
   int x, y;
   Fl_Button *button;
@@ -668,7 +669,9 @@ void OptionsDialog::createCompressionPage(int tx, int ty, int tw, int th)
 void OptionsDialog::createSecurityPage(int tx, int ty, int tw, int th)
 {
 #if defined(HAVE_GNUTLS) || defined(HAVE_NETTLE) || !defined(WIN32)
-  Fl_Group *group = new Fl_Group(tx, ty, tw, th, _("Security"));
+  Fl_Scroll *scroll = new Fl_Scroll(tx, ty, tw, th, _("Security"));
+  tw -= Fl::scrollbar_size();
+  Fl_Group *group = new Fl_Scroll(tx, ty, tw, 0);
 
   int orig_tx;
   int width;
@@ -828,10 +831,15 @@ void OptionsDialog::createSecurityPage(int tx, int ty, int tw, int th)
   tx = orig_tx;
   ty += INNER_MARGIN;
 
-
 #endif // !WIN32
 
+  /* Needed for resize to work sanely */
+  group->resizable(NULL);
+  ty += OUTER_MARGIN - INNER_MARGIN;
+  group->size(group->w(), ty - group->y());
+
   group->end();
+  scroll->end();
 
 #endif // defined(HAVE_GNUTLS) || defined(HAVE_NETTLE) || !defined(WIN32)
 }
```

**Comment by CendioOssman** [comment]
Sorry about the delay. You are not forgotten, we're just a bit too busy at the moment. Please bear with us a bit more, and we will get this merged eventually.

**Comment by CendioOssman** [comment]
The options dialog changes are now merged, so please have a look at rebasing this PR so we can focus on just your new additions.

**Comment by dmcmahill** [comment]
> The options dialog changes are now merged, so please have a look at rebasing this PR so we can focus on just your new additions.

rebased and force pushed

File: vncviewer/OptionsDialog.cxx [hunk 4]
```diff
@@ -786,9 +798,50 @@ void OptionsDialog::createSecurityPage(int tx, int ty, int tw, int th)
   /* Back to normal */
   tx = orig_tx;
   ty += INNER_MARGIN;
+#endif // defined(HAVE_GNUTLS) || defined(HAVE_NETTLE)
+
+#ifndef WIN32
+  /* Connection */
+  ty += GROUP_LABEL_OFFSET;
+  //  height = GROUP_MARGIN * 2 + (INPUT_LABEL_OFFSET + INPUT_HEIGHT) * 1;
+  connectionGroup = new Fl_Group(tx, ty, width, 0, _("Connection"));
```
Comment (CendioOssman) [inline vncviewer/OptionsDialog.cxx:810]: "Connection" is a bit general, so I think we need something more specific here. "Gateway" is what's used in the documentation, so something including that word would probably be best for consistency.

File: vncviewer/OptionsDialog.h [hunk 5]
```diff
@@ -63,6 +63,7 @@ class OptionsDialog : public Fl_Window {
 
   static void handleX509(Fl_Widget *widget, void *data);
   static void handleRSAAES(Fl_Widget *widget, void *data);
+  static void handleViaHost(Fl_Widget *widget, void *data);
```
Comment (CendioOssman) [inline vncviewer/OptionsDialog.h:66]: This doesn't seem to be used?

File: vncviewer/parameters.cxx [hunk 6]
```diff
@@ -164,7 +164,7 @@ BoolParameter fullscreenSystemKeys("FullscreenSystemKeys",
                                    true);
 
 #ifndef WIN32
-StringParameter via("via", "Gateway to tunnel via", "");
+StringParameter via("via", "SSH host for tunnel", "");
```
Comment (CendioOssman) [inline vncviewer/parameters.cxx:167]: This is an unrelated change, so it should be a separate commit.

The term "Gateway" is also used elsewhere, so everything would need to be adjusted so things are consistent.

Lastly, I'm uncertain if this is clearer. We're gaining the detail of it being SSH, but losing the gateway¹ aspect.

¹ Which is a less common description than "proxy"

File: vncviewer/vncviewer.cxx [hunk 7]
```diff
@@ -587,9 +587,23 @@ createTunnel(const char *gatewayHost, const char *remoteHost,
...
+
+  /* strip leading whitespace from gatewayHost */
+  while (*tmps != '\0' && isspace(*tmps)) tmps++;
+
+  /* strip trailing whitespace from gatewayHost */
+  tmps2 = tmps;
+  while (*tmps2 != '\0' && !isspace(*tmps2)) tmps2++;
+  if (isspace (*tmps2)) *tmps2 = '\0';
+  vlog.info(_("Establish SSH tunnel to gatewayHost \"%s\"\n"), tmps);
+
```
Comment (CendioOssman) [inline vncviewer/vncviewer.cxx:602]: These are nice improvements, but unrelated to adding the GUI. So please have this as a separate commit.

File: vncviewer/vncviewer.cxx [hunk 8]
```diff
@@ -809,8 +823,24 @@ int main(int argc, char** argv)
...
+      tmps = via.getValueStr();
+      for( size_t i=0; i < strlen(tmps) ; i++) {
+        if (! isspace(tmps[i]) ) {
+          useSSH = true;
+        }
+      }
+      if (useSSH && mktunnel() != 0)
+        usage(argv[0]);
+      strFree(tmps);
+    }
```
Comment (CendioOssman) [inline vncviewer/vncviewer.cxx:843]: Same here.

File: vncviewer/OptionsDialog.cxx [hunk 9]
```diff
@@ -63,7 +64,7 @@ std::map<OptionsCallback*, void*> OptionsDialog::callbacks;
 static std::set<OptionsDialog *> instances;
 
 OptionsDialog::OptionsDialog()
-  : Fl_Window(580, 480, _("TigerVNC Options"))
+  : Fl_Window(580, 420, _("TigerVNC Options"))
```
Comment (CendioOssman) [inline vncviewer/OptionsDialog.cxx:67]: Please include this commit into the main one so we can avoid this back and forth.

**Review by CendioOssman — CHANGES_REQUESTED** [review]
Great. I have some minor things that need adjusting, then we can go ahead and merge this.

File: vncviewer/OptionsDialog.h (hunk 5 above)
Comment (dmcmahill) [reply vncviewer/OptionsDialog.h:66]: Thanks for catching this.  It was there in the original version before your new options work and when I reworked the code I accidentally left this in.   I'll remove it.

File: vncviewer/parameters.cxx (hunk 6 above)
Comment (dmcmahill) [reply vncviewer/parameters.cxx:167]: I'll remove this particular change.

File: vncviewer/vncviewer.cxx (hunk 7 above)
Comment (CendioOssman) [reply vncviewer/vncviewer.cxx:602]: Yes, that's precisely right. :)

**Comment by vejnar** [comment]
This would be a fantastic feature! Thanks for working on it. PR seems almost done. Any news @dmcmahill to push for the finish line?

Question: Would this option be available on Windows?

**Comment by CendioOssman** [comment]
Unfortunately, it seems this PR is abandoned. :/
